hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RsqlParserValidationOracle.java

```diff
@@ -109,15 +111,17 @@ public ValidationOracleContext suggest(final String rsqlQuery, final int cursorP
         return Collections.emptyList();
     }
 
-    private static void setExceptionDetails(final Exception ex, final List<SuggestToken> expectedTokens) {
-        expectedTokens.addAll(getNextTokens(ex));
-    }
-
-    private static List<SuggestToken> getNextTokens(final Exception e) {
-        final ParseException parseException = findParseException(e);
+    private static void setExceptionDetails(final String rsqlQuery, final Exception ex,
+            final List<SuggestToken> expectedTokens) {
+        final ParseException parseException = findParseException(ex);
         if (parseException == null) {
-            return Collections.emptyList();
+            expectedTokens.addAll(getComparatorOperatorSuggestions(rsqlQuery));
```

Comment:
Why you're waiting until an exception is occurred to suggest one more token. Is there no way to do it early before the execption is thrown. Maybe within `getLogicalOperatorSuggestion()`?

Reply:
The exception is thrown before the `getLogicalOperatorSuggestion()` method is called. In fact, we get a generic `RSQLParserException` from `cz.jirutka.rsql.parser` package, coming from the `targetManagement.findByRsql()` method with no _cause_ specified. In that case, we search for the "!" or "=" symbols at the end of the string in order to suggest the logical operators after the key for map fields.